src/box/box.cc

```diff
@@ -3726,11 +3732,33 @@ box_process1(struct request *request, box_tuple_t **result)
 				"box.ctl.is_recovery_finished() "
 				"to check that snapshot recovery was completed");
 			diag_log();
-			return -1;
+			goto error;
 		}
 	}
 
-	return box_process_rw(request, space, result);
+	if (box_process_rw(request, space, result) != 0)
+		goto error;
+	return 0;
+error:
+	struct error *e = diag_last_error(diag_get());
+	if (tuple != NULL)
+		error_set_mp(e, "tuple", tuple, tuple_end - tuple);
+	/* space pointer can be invalid at this point. */
+	space = space_by_id(request->space_id);
+	if (space != NULL)
+		error_set_str(e, "space", space->def->name);
+	if (request->index_id != 0) {
+		assert(request->key != NULL);
+		assert(request->key_end != NULL);
+		error_set_mp(e, "key", key, key_end - key);
+		if (space != NULL) {
+			struct index *index = space_index(space,
+							  request->index_id);
+			if (index != NULL)
+				error_set_str(e, "index", index->def->name);
+		}
+	}
+	return -1;
```

Comment:
You can and should set the operation right here - by looking at the request type.